src/main/java/net/minecraftforge/registries/DeferredRegister.java

```diff
@@ -170,6 +171,7 @@ private DeferredRegister(IForgeRegistry<T> reg, String modid)
      * @param name The new entry's name, it will automatically have the modid prefixed.
      * @param sup A factory for the new entry, it should return a new instance every time it is called.
      * @return A RegistryObject that will be updated with when the entries in the registry change.
+     * @apiNote This method will return {@link Holder} in future versions.
      */
     @SuppressWarnings({ "rawtypes", "unchecked" })
     public <I extends T> RegistryObject<I> register(final String name, final Supplier<? extends I> sup)
```

Comment:
This could be made to return Holder right now, and backwards compat can be added via a js coremod.

Reply:
That still breaks source compatibility

Reply:
Yeah but do we care if RegistryObject is getting deleted?

Reply:
You could also theoretically do bytecode manipulation on the universal jar itself before publishing so that it theoretically would still work in IDEs and compile.

Reply:
> Yeah but do we care if RegistryObject is getting deleted?
RO will probably be replaced with a custom Holder implementation is lazy-loaded and delegating to `Holder$Reference` - the question is should this new method return that strongly typed delegating holder, or just `Holder<T>`

Reply:
imo, make a choice:

1) You're deleting `RO`, so don't expose anything else. Return `Holder<T>`.
2) You're returning a custom interface / implementation of `Holder`, `Supplier`, etc. This needs a name, so how about we call this interface `RegistryObject`. Then we obviously return that class.

Personally, I'd prefer option 2, because that means minimal churn on the mod dev side.

Reply:
We can do option 2, but I would like to proceed with a rename to make it obvious that the custom class _is_ a `Holder<T>`.

Given our current scheme, the most apt name is `DeferredHolder`

Reply:
One issue I see with making this not return `RegistryObject` at least at dev time (even if it is added back at runtime via coremod) may cause issues if any devs want to be supporting both Forge and NeoForge as best as they can if they don't require anything that is specific to one version vs the other. As if it doesn't provide a `RegistryObject` at compile time then those devs will be basically forced to compile against Forge rather than being able to just compile against NeoForge.

Reply:
Yeah, leaving it as RO for now. Too annoying to do dual support.